x-pack/plugins/observability_solution/dataset_quality/server/services/privileges.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { forbidden } from '@hapi/boom';
+import type { SecurityIndexPrivilege } from '@elastic/elasticsearch/lib/api/types';
+import type { ElasticsearchClient } from '@kbn/core/server';
+
+import { streamPartsToIndexPattern } from '../../common/utils';
+import { DEFAULT_DATASET_TYPE } from '../../common/constants';
+
+class DatasetQualityPrivileges {
+  public async getHasIndexPrivileges(
+    esClient: ElasticsearchClient,
+    indexes: string[],
+    privileges: SecurityIndexPrivilege[]
+  ): Promise<Awaited<Record<string, boolean>>> {
+    const indexPrivileges = await esClient.security.hasPrivileges({
+      index: indexes.map((dataStream) => ({ names: dataStream, privileges })),
+    });
+
+    const indexesList = Object.keys(indexPrivileges.index);
+    return indexesList.reduce((acc, index) => {
+      const privilegesList = Object.values(indexPrivileges.index[index]);
+      const hasAllPrivileges = privilegesList.every((hasPrivilege) => hasPrivilege);
+
+      return Object.assign(acc, { [index]: hasAllPrivileges });
+    }, {} as Record<string, boolean>);
```

Comment:
Nit: can we pass this as the generic argument to `reduce` rather than casting here?